user_topic_popover: Use new sliding tab for topic visibility switcher.

sayamsamal opened the pull request:
This PR adds the new sliding tab animation to the topic visibility switchers (one situated in the message header topic visibility popover, and the other one situated in the sidebar topic actions), along with some functionality changes, such as:
- Hiding the popover, only when the mute topic request gets completed successfully.
- Adding visual feedback via the same sliding animation, on request failure.

Fixes: [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20theme.20switcher.20.2322803/near/1766900)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Light Mode</summary>

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/82862779/254813a0-bdc4-423e-97fc-09565314a6ef) | ![image](https://github.com/zulip/zulip/assets/82862779/6ef10c85-d9e5-40ec-a358-4dbe069f0fd2) | 

</details>

<details>
<summary>Light Mode (Sidebar topic actions popover)</summary>

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/82862779/b49ed3e2-5d6f-4b41-adb0-8f9008c65b83) | ![image](https://github.com/zulip/zulip/assets/82862779/5040db02-321f-4b26-a987-15edc3d9b645) | 

</details>

<details>
<summary>Dark Mode</summary>

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/82862779/5a44892a-84ca-4ae2-bde6-8a1984840bdb) | ![image](https://github.com/zulip/zulip/assets/82862779/bf6aab51-d3b2-44ee-b196-a95cbe19e670) | 

</details>

<details>
<summary>Dark Mode (Sidebar topic actions popover)</summary>

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/82862779/3b17c702-d2f2-49fa-a6d2-c1a089c396a1) | ![image](https://github.com/zulip/zulip/assets/82862779/2827fb45-64c1-414e-992b-efeb43c4b2ff) | 

</details>

<details>
<summary>General Usage</summary>

[general-usage.webm](https://github.com/zulip/zulip/assets/82862779/6ad998bd-c487-4ed9-b3ce-43dbe22a05ef)

[topic-vis-popover-working.webm](https://github.com/zulip/zulip/assets/82862779/39185559-30f8-4b8c-9075-e78bedc0df4c)


</details>

<details>
<summary>Request error feedback</summary>

[sidebar-topic-vis-error-handling.webm](https://github.com/zulip/zulip/assets/82862779/2265ad0d-8683-442e-b436-4c6ad549e23e)

[error-handling-topic-vis-popover.webm](https://github.com/zulip/zulip/assets/82862779/3a6d5ebb-170e-4f04-ace9-2692bd37061d)

</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Hm, this doesn't quite feel right in my manual testing.

I think we should keep the current behavior, where the menu automatically closes after you make the selection. In this PR, the menus that just have the topic visibility switcher remain open.

alya commented:
With the addition of this animation, I'd also like to see the left sidebar menu (and the other ones, I would guess) remain open a bit longer, so that you can really visually process that your change took place.

alya commented:
The visibility popovers feel pretty good to me now. In left sidebar topic menu, it still feels a bit quick, maybe because there's more going on in that menu, so it's harder to focus on a particular part of the UI.

sayamsamal commented:
## Updates:

- Addressed (https://github.com/zulip/zulip/pull/29479#issuecomment-2030606390). The issue was related to the updation of the reference elements, which caused the popovers to force close.
- Removed redundant code from 3f2ab44f94.
- Other code cleanups in general and also found and removed some of my own bugs.

sayamsamal commented:
@prakhar1144 Could you take a look at commit 10184554b443e1bd5ea23de4c0fd020d0acf01a7? It is directly related to your work in 3f2ab44f9485da603dc192b1b6a03102d13997d4. I would really appreciate your insight into this, thanks!

prakhar1144 commented:
> @prakhar1144 Could you take a look at commit [1018455](https://github.com/zulip/zulip/commit/10184554b443e1bd5ea23de4c0fd020d0acf01a7)? It is directly related to your work in [3f2ab44](https://github.com/zulip/zulip/commit/3f2ab44f9485da603dc192b1b6a03102d13997d4). I would really appreciate your insight into this, thanks!

Yeah, that's right. Thanks for cleaning it up.

alya commented:
Looks good to me in manual testing!

timabbott commented on web/src/user_topic_popover.js:
```diff
@@ -59,6 +58,18 @@ export function initialize() {
+                            .parent()
+                            .find(`input[data-visibility-policy="${prev_visibility_policy}"]`)
+                            .prop("checked", true);
+                    }, 500);
```
What's the motivation for the timeout here?

sayamsamal replied:
Thank you for the insight, this did not cross my mind but makes a lot of sense. I have added a utility function `util.get_remaining_time` which calculates the remaining time for delay based on the initial time of the toggle being clicked.

The basic principle of the function is:
- If RTT < Required delay time,
Wait for (Delay - RTT) and then perform the action.

- If RTT >=  Required delay time,
Immediately perform the action, since the minimum required time for the animation to play through has already elapsed.

timabbott replied:
Sounds good!

timabbott commented on web/src/user_topics_ui.js:
```diff
@@ -13,11 +12,18 @@ import * as user_topics from "./user_topics";
+        stream_list.update_streams_sidebar();
+        unread_ui.update_unread_counts();
+        message_lists.current?.update_muting_and_rerender();
+    }, 500);
```
This delay seems like it could be conditional on the popover being open? I don't see a reason to defer it in other cases.

sayamsamal replied:
The reason for adding this delay here is that these methods are responsible for hiding/updating the reference element of the popover when a topic is muted. This leads to the force closing of the popovers since their reference element is no longer in the view. Thus, we are left with half-fulfilled animations.

For more clarity, I have separated this into its own commit (d60fb05416c4f8ae2935d84625281f4fab6f1468), added a better comment on the "why" and restricted this delay to only come into action when a topic is muted.

sayamsamal replied:
I have also included `recent_view_ui.update_topic_visibility_policy` into this delay group since that also removed the topic from view when muted. I missed this the last time.

timabbott commented on web/src/topic_popover.js:
```diff
@@ -98,42 +98,6 @@ export function initialize() {
-                        user_topics.all_visibility_policies.UNMUTED,
-                    );
-                    popover_menus.hide_current_popover_if_visible(instance);
-                });
```
Just to be clear, this code is now handled by `$popper.on("change", "input[name='sidebar-topic-visibility-select']"`, right?

sayamsamal replied:
Yes, that's correct!

timabbott commented:
Looks good to me aside from the few notes above about delays that might not be scoped correctly. I like how much code this deletes!

timabbott commented:
Looks good to me, except for the last commit, we should only do the delay if in fact the popover was open.

sayamsamal commented:
@timabbott I have updated the last commit to only apply delay where the forced closing of the popover is expected. The logic I used is:

Apply delay only if, `is_topic_muted && is_relevant_popover_open && !is_inbox_view && !is_topic_narrow`.

Where relevant popovers are: The topic action menu popover + The message header's visibility policy popover. And, the inbox view and the topics narrow view don't remove the topic out of the view.

sayamsamal commented:
Did some manual testing as well:
- Inbox view, delay expected? False. 
![inbox_view](https://github.com/zulip/zulip/assets/82862779/a420bc14-dc67-46d5-8eef-ea5a5ae1db8a)
- Topics Narrow, delay expected? False.
![topics_narrow](https://github.com/zulip/zulip/assets/82862779/50d7d341-8b68-40eb-9f5b-6010a15fcafd)
- Streams Narrow, delay expected? True.
![streams_narrow](https://github.com/zulip/zulip/assets/82862779/fbbf7247-d371-4e51-80df-48a392fb2b0a)

timabbott commented:
Looks great, merged, thanks @sayamsamal!
